crates/sui-types/src/intent.rs

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::batch::AuthorityBatch;
+use crate::messages::{TransactionData, TransactionEffects};
+use crate::messages_checkpoint::CheckpointSummary;
+
+use serde::{Deserialize, Serialize};
+
+/// Current Sui version (for compatibility purposes).
+/// TODO: Increment value when a non-backwards compatible Sui version is launched (i.e. when
+/// changing the format, signature or field ordering of signable types).
+const SUI_COMPATIBILITY_VERSION: u8 = 0;
+const SUI_CHAIN_ID: u8 = 0;
+
+/// This is the only type we should sign per our serialization-handbook to provide domain separation
+/// and avoid accidental serialized-bytes collisions between structs and intents, different versions
+/// of the same struct and chains (mainnet, testnet etc).
+/// Similarly, `SecureIntent` should be preferred when hashing requires domain separation
+/// guarantees.
+pub trait SecureIntent: Serialize + serde::de::DeserializeOwned + private::SealedIntent {}
+
+/// Struct required when Sui repository-version and chainID are required for domain
+/// separation (i.e. when something is sign-able).
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
+pub struct VersionPlusChainId {
+    version: u8,
+    chain_id: u8,
+}
+
+impl Default for VersionPlusChainId {
+    // TODO: read these values from some config, genesis or epoch.
+    fn default() -> Self {
+        VersionPlusChainId {
+            version: SUI_COMPATIBILITY_VERSION,
+            chain_id: SUI_CHAIN_ID,
+        }
+    }
+}
+
+/// `IntentScope` is required to guarantee two different intents will never collide.
+pub enum IntentScope {
+    TransactionData,
+    TransactionEffects,
+    AuthorityBatch,
+    CheckpointSummary,
+    PersonalMessage,
+}
+
+impl IntentScope {
+    /// Specifically assign a byte per enum element, to avoid accidental issues (i.e., field
+    /// swapping, adding in the middle or deletion) which would affect enum's BCS serialization.
+    ///
+    /// IMPORTANT NOTE: we should ensure forward and backward uniqueness of scope-values. That said,
+    /// if a value has been used in the past, then it should NEVER be reassigned to a different
+    /// scope.
+    pub fn value(&self) -> u8 {
+        match *self {
+            IntentScope::TransactionData => 0,
+            IntentScope::TransactionEffects => 1,
+            IntentScope::AuthorityBatch => 2,
+            IntentScope::CheckpointSummary => 3,
+            IntentScope::PersonalMessage => 4,
+        }
+    }
+}
+
+// --- TransactionData intent ---
+
+/// The intent we should use for TransactionData signing.
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
+pub struct TransactionDataIntent {
+    version_plus_chain_id: VersionPlusChainId,
+    scope: u8,
+    signable: TransactionData,
+}
```

Comment:
I'm not sure how i feel about these distinct "intent" structs. They definitely are ridged and enforce what can be signed but i'm not sure how much that actually buys us. I think having something a bit more flexible like in patch I've linked may make a bit more sense.

Reply:
I wanted to avoid mixing intents with non-related object types, so one way was to wrap objects into intents. That would still allow defining multiple wrappers for the same object type, just create an extra Intent that wraps the same type.